pkg/apis/apps/validation/validation.go

```diff
@@ -69,6 +69,26 @@ func ValidatePodTemplateSpecForStatefulSet(template *api.PodTemplateSpec, select
 	return allErrs
 }
 
+func ValidatePersistentVolumeClaimRetentionPolicyType(policy apps.PersistentVolumeClaimRetentionPolicyType, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	switch policy {
+	case apps.RetainPersistentVolumeClaimRetentionPolicyType:
+	case apps.DeletePersistentVolumeClaimRetentionPolicyType:
+	default:
+		allErrs = append(allErrs, field.NotSupported(fldPath, policy, []string{string(apps.RetainPersistentVolumeClaimRetentionPolicyType), string(apps.DeletePersistentVolumeClaimRetentionPolicyType)}))
+	}
+	return allErrs
+}
+
+func ValidatePersistentVolumeClaimRetentionPolicy(policy *apps.StatefulSetPersistentVolumeClaimRetentionPolicy, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
```

Comment:
do we need the feature gate check here? If feature gate is not enabled, validation should fail for any policy set here?

Reply:
My understanding is that we don't want the feature gate check, otherwise if you disable the feature gate on an existing cluster, resources with the policy set will fail validation and you'll be unable to do anything with them.

Reply:
Yes, we do need the feature gate.  If the gate is off, we don't validate the fields.  If the gate is on, we validate.

This pairs with the "drop unset fields" logic which clears set fields when the gate is off prior to write.

Reply:
Even if the feature gate is off, in a rollback scenario, the fields can still exist. In our api-changes doc, we do not validate based on feature gates: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-field-in-existing-api-version

Reply:
I originally did validate, but in the last review was convinced to remove validation: https://github.com/kubernetes/kubernetes/pull/99378#discussion_r596990796

I'm not really sure about this process where the implementation keeps changing depending on who is reviewing :-/

Reply:
Unfortunately, the other reviewer just missed that this is a project wide enforced thing (we have a set of rules for how alpha fields are treated) and we want to apply it consistently.  I.e. looking through the other statefulset gated fields they all obey the "only validate the field if gate enabled"

Reply:
Hold up on the validation change, the other fields are wrong and we didn't clarify why.  https://github.com/kubernetes/community/pull/6236 will try to correct that, but for right now you should not have validation within feature gate blocks.